README.md
# oeg.infotech.tool

[![Python 2.7](https://img.shields.io/travis/vb64/oeg.infotech.tool.svg?label=Python%202.7&style=plastic)](https://travis-ci.org/vb64/oeg.infotech.tool)
[![Code Climate](https://img.shields.io/codeclimate/maintainability-percentage/vb64/oeg.infotech.tool.svg?label=Code%20Climate&style=plastic)](https://codeclimate.com/github/vb64/oeg.infotech.tool)
[![Coverage Status](https://coveralls.io/repos/github/vb64/oeg.infotech.tool/badge.svg?branch=master)](https://coveralls.io/github/vb64/oeg.infotech.tool?branch=master)

Win32 console utility for manipulating OrgEnergoGaz Infotech xml files

## Install

```
$ git clone git@github.com:vb64/oeg.infotech.tool.git
$ cd oeg.infotech.tool
$ make setup PYTHON_BIN=D:\python\2.7.15\python.exe
```

## Build infotech.exe

```
$ make exe
```

`infotech.exe` executable can be found in `dist` folder.

## Use

```
> infotech.exe csv_weld fixtures\xml\1736.xml >build\1736_weld.csv
> infotech.exe csv_defect fixtures\xml\1736.xml >build\1736_defect.csv
> infotech.exe csv_lineobj fixtures\xml\1736.xml >build\1736_lineobj.csv
> infotech.exe reverse fixtures\xml\1736.xml >build\1736_reverse.xml
> infotech.exe join fixtures\xml\1736.xml 1100 fixtures\xml\1737.xml >build\1736_join.xml
```


history.txt
12.02.2020 infotech.exe ver.1.5
-------------------------------

+ поддержка полей "ИУСТ" в `infotech.xml` (версия 1.4.1 [библиотеки oeg-infotech](https://github.com/vb64/oeg.infotech.xml)).

* автотесты переведены на использование pytest

15.03.2019 infotech.exe ver.1.4
-------------------------------

* программа переведена на использование библиотеки https://github.com/vb64/oeg.infotech.xml

03.10.2018 infotech.exe ver.1.3
-------------------------------

* программа упакована в один exe файл без дополнительных модулей.

27.09.2018 infotech.exe ver.1.2
-------------------------------

* кодировка создаваемого xml файла изменена с "cp1251" на "windows-1251"

10.09.2018 infotech.exe ver.1.1
-------------------------------

* при выводе содержимого xml файла в csv файл, в полях, содержащих значения с плавающей точкой (ориентация и толщина), символ десятичной точки заменяется на символ запятой. при открытии кликом по такому csv файлу MS Excel не преобразовывает значение таких полей в дату.

10.09.2018 infotech.exe ver.1.0
-------------------------------

Утилита командной строки infotech.exe предназначена для обработки .xml файлов ИНФОТЕХ, созданных из базы данных PARC. Использование:

> infotech.exe команда файл.xml

Команды csv_weld, csv_lineobj, csv_defect выводят на экран содержимое секций WELDS, LINEOBJS, DEFECTS (соответствено) .xml файла в виде .CSV файла, котороый можно открыть в MS Excel.
Для записи результата работы программы в файл нужно использовать символ перенаправления вывода '>'. Например, для создания файла швы.csv, содержащего все швы из файла инфотех.xml, нужно использовать следующую команду:

> infotech.exe csv_weld инфотех.xml >швы.csv

Команда reverse "переворачивает" последовательность швов, дефектов и линейных обьектов в .xml файле и выводит на экран содержимое "перевернутого" xml файла. Данные xml файла при этом преобразуются следующим образом:

- Дистанции швов и линейных обьктов пересчитываются с конца записи (дистанция последнего шва + длина последней трубы)
- Дистанции дефектов пересчитываются с конца записи с учетом длины дефекта (конец дефекта становится его началом)
- Ориентации дефектов и продольных швов пересчитываются в зеркальном отображении (ориентация на 3 часа становится ориентацией на 9 часов и т.п.)
- Существующая нумерация труб и дефектов сбрасывается, трубы и дефекты нумеруются в соответствии с новым порядком, новая нумерация начинается с 1.
- У линейных обьектов, имеющих начало и конец, эти пары меняются местами (начало патрона становится концом патрона, конец патрона становится началом патрона). Такая перестановка проводится для следующих пар линейных обьектов:
"Начало водной преграды", "Конец водной преграды"
"Начало дистанции", "Конец дистанции"
"Нарушения адгезии изоляции трубы начало", "Нарушения адгезии изоляции трубы конец"
"Отвод (поворот) начало", "Отвод (поворот) конец"
"Патрон начало", "Патрон конец"
"Сегментный участок начало", "Сегментный участок конец"
"Участок пригрузов начало", "Участок пригрузов конец"

Например, для создания "перевернутого" файла revert.xml из файла инфотех.xml нужно использовать следующую команду:

> infotech.exe reverse инфотех.xml >revert.xml

Команда fix исправляет неверные данные снаряда "УМДП-1400" в секции PIGPASS, если ссылка на данный снаряд имеется в этой секции.
Исправляются следующие поля элемента секции PIGPASS: MANUFACT_DATE="2017" PIGTYPE="990004033563" OBSLTYPE="2"

Например, для создания исправленной версии файла инфотех.xml и записи ее в файл fixed.xml нужно использовать следующую команду:

> infotech.exe fix инфотех.xml >fixed.xml

Команда join обьединяет несколько xml файлов в один файл. В качестве заголовка итогового файла и данных о пропусках (секция PIGPASS) выбираются данные из первого файла в цепочке.
Дистанции всех обьектов пересчитываются в соответствии с их расположением в цепочке файлов. Нумерация труб и дефектов начинается с 1.
В цепочке файлов могут присутствовать "стыковочные" элементы.
Если в цепочке файлов есть элементы, состоящие из одних цифр, то такой элемент интерпретируется как труба типа "Неопределенная секция" длиной в указанное кол-во сантиметров, без продольных швов и с толщиной, равной толщине предыдущей трубы.

Например, чтобы пристыковать к обьектам файла первый.xml две трубы длинами 11 и 10 метров, а затем данные из файла второй.xml и записать полученный результат в файл результат.xml, нужно использовать следующую команду:

> infotech.exe join первый.xml 1100 1000 второй.xml >результат.xml


tests/conftest.py
"""
Module for environment setup and tests runner
"""
import os
import sys


def path_setup():
    test_dir = os.path.dirname(os.path.abspath(__file__))
    # sys.path.insert(1, test_dir)
    sys.path.insert(1, os.path.join(os.path.dirname(test_dir), 'source'))


path_setup()


tests/test/test_console.py
"""
make test T=test_console.py
"""
from . import TestInfotech


class TestConsole(TestInfotech):
    """
    console
    """
    def test_console(self):
        """
        console
        """
        from infotech import main, Commands

        self.assertEqual(main([]), 1)
        self.assertEqual(main(['', 'wrong_command', 'wrong_file']), 1)
        self.assertEqual(main(['', Commands.REVERSE, 'wrong_file']), 1)
        self.assertEqual(main(['', Commands.REVERSE, self.fixture('1736.xml')]), 0)
        self.assertEqual(main(['', Commands.CSV_WELD, self.fixture('1736.xml')]), 0)
        self.assertEqual(main(['', Commands.CSV_DEFECT, self.fixture('1736.xml')]), 0)
        self.assertEqual(main(['', Commands.CSV_LINEOBJ, self.fixture('1736.xml')]), 0)
        self.assertEqual(main(['', Commands.REVERSE, self.fixture('1736.xml')]), 0)
        self.assertEqual(main(['', Commands.FIX, self.fixture('1736.xml')]), 0)
        self.assertEqual(main(['', Commands.FIX, self.fixture('umdp-1400.xml')]), 0)
        self.assertEqual(main(['', Commands.JOIN, self.fixture('1736.xml'), '1100', self.fixture('1737.xml')]), 0)

        self.assertEqual(main(['', Commands.MAKEMONEY, self.fixture('1736.xml')]), 0)
        self.assertEqual(main(['', Commands.CSV_DEFECT, self.fixture('navigation.xml'), True]), 0)


source/infotech.py
"""
console client for infotech xml files processing
"""
import sys
from oeg_infotech import Infotech

VERSION = '1.5'
COPYRIGHTS = 'Copyrights by Vitaly Bogomolov 2018-2020'


class Commands:  # pylint: disable=old-style-class,no-init,too-few-public-methods
    """
    available commands
    """
    MAKEMONEY = '_make_me_rich'
    REVERSE = 'reverse'
    CSV_WELD = 'csv_weld'
    CSV_DEFECT = 'csv_defect'
    CSV_LINEOBJ = 'csv_lineobj'
    FIX = 'fix'
    JOIN = 'join'


def get_navigation(argv):
    """
    return with_navigation flag from arg list
    """
    if len(argv) > 3:
        return bool(argv[3])

    return False


def cmd_csv_welds(info, argv):
    """
    dump WELDS section as csv
    """
    print info.welds.as_csv(with_navigation=get_navigation(argv))
    return 0


def cmd_csv_defects(info, argv):
    """
    dump DEFECTS section as csv
    """
    print info.defects.as_csv(with_navigation=get_navigation(argv))
    return 0


def cmd_csv_lineobjects(info, argv):
    """
    dump LINEOBJS section as csv
    """
    print info.lineobjects.as_csv(with_navigation=get_navigation(argv))
    return 0


def cmd_reverse(info, _argv):
    """
    reverse vector of objects in xml and dump updated xml to string
    """
    print info.reverse()
    return 0


def cmd_fix(info, _argv):
    """
    repair umdp-1400 data
    """
    print info.fix()
    return 0


def cmd_join(info, argv):
    """
    join several xml files and connect tubes
    """
    print info.join(argv[3:])
    return 0


def cmd_stub(_info, _argv):
    """
    stub for fake/new command
    """
    print "Not implemented yet"
    return 0


COMMANDS = {
  Commands.MAKEMONEY: cmd_stub,
  Commands.REVERSE: cmd_reverse,
  Commands.CSV_WELD: cmd_csv_welds,
  Commands.CSV_DEFECT: cmd_csv_defects,
  Commands.CSV_LINEOBJ: cmd_csv_lineobjects,
  Commands.FIX: cmd_fix,
  Commands.JOIN: cmd_join,
}


def main(argv):
    """
    entry point
    """
    cmd_list = sorted(COMMANDS.keys()[1:])

    if len(argv) < 3:
        print "Infotech xml file processor. Version {} {}".format(VERSION, COPYRIGHTS)
        print "Usage:\ninfotech command infotech_file.xml\n"
        print "Commands:"
        print "csv_weld - print WELDS section of .xml file to stdout as .CSV file"
        print "csv_lineobj - print LINEOBJS section of .xml file to stdout as .CSV file"
        print "csv_defect - print DEFECTS section of .xml file to stdout as .CSV file"
        print "reverse - revert vector of objects in .xml file and print result as .XML file to stdout"
        print "join file1.xml 1100 file2.xml - print outer joined result xml file"

        return 1

    if argv[1] not in COMMANDS:
        print "Unknown command: {}\nAvailable commands: {}".format(argv[1], ' '.join(cmd_list))
        return 1

    try:
        info = Infotech.from_file(argv[2])
    except IOError as exc:
        print exc
        return 1

    command = COMMANDS[argv[1]]
    return command(info, argv)


if __name__ == '__main__':  # pragma: no cover
    sys.exit(main(sys.argv))
